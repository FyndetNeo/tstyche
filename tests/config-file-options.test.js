import { afterEach, describe, expect, test } from "@jest/globals";
import { clearFixture, writeFixture } from "./__utils__/fixtureFactory.js";
import { normalizeOutput } from "./__utils__/normalizeOutput.js";
import { spawnTyche } from "./__utils__/spawnTyche.js";

const isStringTestText = `import { expect, test } from "tstyche";
test("is string?", () => {
  expect<string>().type.toBeString();
});
`;

const isNumberTestText = `import { expect, test } from "tstyche";
test("is number?", () => {
  expect<number>().type.toBeNumber();
});
`;

const tsconfig = {
  extends: "../tsconfig.json",
  include: ["**/*"],
};

const fixture = "config-file-options";

afterEach(async () => {
  await clearFixture(fixture);
});

describe("tstyche.config.json", () => {
  test("'target' option", async () => {
    const config = {
      target: ["4.8", "5.3.2", "current"],
    };

    await writeFixture(fixture, {
      ["__typetests__/dummy.test.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      ["tstyche.config.json"]: JSON.stringify(config, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("handles 'current' as 'target' option value", async () => {
    const config = {
      target: ["current"],
    };

    await writeFixture(fixture, {
      ["__typetests__/dummy.test.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      ["tstyche.config.json"]: JSON.stringify(config, null, 2),
    });

    const { exitCode, stderr } = await spawnTyche(fixture);

    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("'testFileMatch' option", async () => {
    const config = {
      testFileMatch: ["**/type-tests/*.test.ts"],
    };

    await writeFixture(fixture, {
      ["__typetests__/isNumber.test.ts"]: isNumberTestText,
      ["__typetests__/isString.test.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      ["tstyche.config.json"]: JSON.stringify(config, null, 2),
      ["type-tests/isNumber.test.ts"]: isNumberTestText,
      ["type-tests/isString.test.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });
});
