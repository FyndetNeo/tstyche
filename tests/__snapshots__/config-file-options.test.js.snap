// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tstyche.config.json 'target' option: stdout 1`] = `
"adds TypeScript 4.8.4 to <<cwd>>/tests/__fixtures__/config-file-options/.store/4.8.4
uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/dummy.test.ts
  + is string?

adds TypeScript 5.3.2 to <<cwd>>/tests/__fixtures__/config-file-options/.store/5.3.2
uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/dummy.test.ts
  + is string?

uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/dummy.test.ts
  + is string?

Targets:    3 passed, 3 total
Test files: 3 passed, 3 total
Tests:      3 passed, 3 total
Assertions: 3 passed, 3 total
Duration:   <<timestamp>>

Ran all test files.
"
`;
